File: static/js/pages/masternode.js
import { api } from '../api.js';
import { mnStatusBadge, mnTypeBadge, formatNumber, hashLink, fullAddressLink, blockLink, showLoading } from '../components.js';

export async function renderMasternode({ hash }) {
    const app = document.getElementById('app');
    showLoading(app);

    try {
        const mn = await api.get(`/api/masternode/${hash}`);

        const revReasons = { 0: 'Not revoked', 1: 'Compromised', 2: 'Key change', 3: 'Upgrade' };
        const status = mn.pose_ban_height > 0 ? 'POSE_BANNED' : 'ENABLED';

        app.innerHTML = `
            <div class="breadcrumb"><a href="#/masternodes">Masternodes</a> / Detail</div>
            <div class="page-title">
                <h1>Masternode</h1>
                ${mnTypeBadge(mn.mn_type)}
                ${mnStatusBadge(status)}
            </div>

            <div class="card" style="margin-bottom:1.5rem">
                <div class="card-header"><h3>Registration</h3></div>
                <div class="card-body">
                    <div class="detail-grid">
                        <div class="detail-label">ProTx Hash</div>
                        <div class="detail-value">${mn.pro_tx_hash}</div>
                        <div class="detail-label">Type</div>
                        <div class="detail-value normal-font">${mn.mn_type}${mn.mn_type === 'Evo' ? ' (HPMN - 4000 DASH)' : ' (1000 DASH)'}</div>
                        <div class="detail-label">Collateral TX</div>
                        <div class="detail-value">${hashLink(mn.collateral_hash)}:${mn.collateral_index}</div>
                        <div class="detail-label">Collateral Addr</div>
                        <div class="detail-value">${fullAddressLink(mn.collateral_address)}</div>
                        <div class="detail-label">Registered</div>
                        <div class="detail-value">${blockLink(mn.registered_height)}</div>
                        <div class="detail-label">Confirmations</div>
                        <div class="detail-value">${formatNumber(mn.confirmations)}</div>
                        <div class="detail-label">Operator Reward</div>
                        <div class="detail-value">${mn.operator_reward}%</div>
                    </div>
                </div>
            </div>

            <div class="card" style="margin-bottom:1.5rem">
                <div class="card-header"><h3>Network & Service</h3></div>
                <div class="card-body">
                    <div class="detail-grid">
                        <div class="detail-label">Service</div>
                        <div class="detail-value">${mn.service}</div>
                        <div class="detail-label">Owner Address</div>
                        <div class="detail-value">${fullAddressLink(mn.owner_address)}</div>
                        <div class="detail-label">Voting Address</div>
                        <div class="detail-value">${fullAddressLink(mn.voting_address)}</div>
                        <div class="detail-label">Payout Address</div>
                        <div class="detail-value">${fullAddressLink(mn.payout_address)}</div>
                        <div class="detail-label">Operator Key</div>
                        <div class="detail-value" style="font-size:0.85rem">${mn.pub_key_operator}</div>
                        ${mn.platform_node_id ? `
                            <div class="detail-label">Platform Node ID</div>
                            <div class="detail-value">${mn.platform_node_id}</div>
                            <div class="detail-label">Platform HTTP</div>
                            <div class="detail-value">Port ${mn.platform_http_port}</div>
                            <div class="detail-label">Platform P2P</div>
                            <div class="detail-value">Port ${mn.platform_p2p_port}</div>
                        ` : ''}
                    </div>
                </div>
            </div>

            <div class="card" style="margin-bottom:1.5rem">
                <div class="card-header"><h3>Proof of Service</h3></div>
                <div class="card-body">
                    <div class="detail-grid">
                        <div class="detail-label">PoSe Penalty</div>
                        <div class="detail-value ${mn.pose_penalty > 0 ? 'text-warning' : ''}">${mn.pose_penalty}</div>
                        <div class="detail-label">PoSe Ban Height</div>
                        <div class="detail-value ${mn.pose_ban_height > 0 ? 'text-danger' : ''}">${mn.pose_ban_height > 0 ? blockLink(mn.pose_ban_height) : 'Not banned'}</div>
                        <div class="detail-label">PoSe Revived</div>
                        <div class="detail-value">${mn.pose_revived_height > 0 ? blockLink(mn.pose_revived_height) : 'N/A'}</div>
                        <div class="detail-label">Revocation</div>
                        <div class="detail-value normal-font">${revReasons[mn.revocation_reason] || 'Unknown'}</div>
                        <div class="detail-label">Platform Banned</div>
                        <div class="detail-value normal-font">${mn.is_platform_banned ? '<span class="text-danger">Yes</span>' : 'No'}</div>
                    </div>
                </div>
            </div>

            <div class="card">
                <div class="card-header"><h3>Payment & Mixing</h3></div>
                <div class="card-body">
                    <div class="detail-grid">
                        <div class="detail-label">Last Paid</div>
                        <div class="detail-value">${mn.last_paid_height > 0 ? blockLink(mn.last_paid_height) : 'Never'}</div>
                        <div class="detail-label">Consecutive Pays</div>
                        <div class="detail-value">${mn.consecutive_payments}</div>
                        <div class="detail-label">Mixing TX Count</div>
                        <div class="detail-value">${mn.mixing_tx_count}</div>
                        <div class="detail-label">Last DSQ</div>
                        <div class="detail-value">${mn.last_dsq}</div>
                    </div>
                </div>
            </div>
        `;
    } catch (e) {
        app.innerHTML = `<div class="error-message"><h2>Error</h2><p>${e.message}</p></div>`;
    }
}

File: src/models/masternode.rs
use serde::Serialize;

use crate::rpc::types::{RpcMasternodeListEntry, RpcProtx};

#[derive(Debug, Serialize, Clone)]
pub struct MasternodeSummary {
    pub pro_tx_hash: String,
    pub mn_type: String,
    pub status: String,
    pub service: String,
    pub pose_penalty: u32,
    pub last_paid_block: u64,
    pub last_paid_time: u64,
    pub registered_height: Option<u64>,
    pub collateral_address: String,
    pub payout_address: String,
    pub platform_node_id: Option<String>,
    pub platform_http_port: Option<u16>,
    pub platform_p2p_port: Option<u16>,
}

#[derive(Debug, Serialize, Clone)]
pub struct MasternodeDetail {
    pub pro_tx_hash: String,
    pub mn_type: String,
    pub collateral_hash: String,
    pub collateral_index: u32,
    pub collateral_address: String,
    pub operator_reward: f64,
    pub service: String,
    pub registered_height: u64,
    pub last_paid_height: u64,
    pub consecutive_payments: u32,
    pub pose_penalty: u32,
    pub pose_revived_height: i64,
    pub pose_ban_height: i64,
    pub revocation_reason: u32,
    pub owner_address: String,
    pub voting_address: String,
    pub payout_address: String,
    pub pub_key_operator: String,
    pub platform_node_id: Option<String>,
    pub platform_http_port: Option<u16>,
    pub platform_p2p_port: Option<u16>,
    pub is_platform_banned: Option<bool>,
    pub confirmations: i64,
    pub last_dsq: i64,
    pub mixing_tx_count: u32,
}

impl MasternodeSummary {
    pub fn from_list_entry(entry: &RpcMasternodeListEntry) -> Self {
        Self {
            pro_tx_hash: entry.pro_tx_hash.clone(),
            mn_type: entry.mn_type.clone(),
            status: entry.status.clone(),
            service: entry.address.clone(),
            pose_penalty: entry.pose_penalty_score,
            last_paid_block: entry.lastpaidblock,
            last_paid_time: entry.lastpaidtime,
            registered_height: None,
            collateral_address: entry.collateraladdress.clone(),
            payout_address: entry.payee.clone(),
            platform_node_id: entry.platform_node_id.clone(),
            platform_http_port: entry.platform_http_port,
            platform_p2p_port: entry.platform_p2p_port,
        }
    }
}

impl MasternodeDetail {
    pub fn from_protx(protx: &RpcProtx) -> Self {
        Self {
            pro_tx_hash: protx.pro_tx_hash.clone(),
            mn_type: protx.mn_type.clone(),
            collateral_hash: protx.collateral_hash.clone(),
            collateral_index: protx.collateral_index,
            collateral_address: protx.collateral_address.clone(),
            operator_reward: protx.operator_reward,
            service: protx.state.service.clone(),
            registered_height: protx.state.registered_height,
            last_paid_height: protx.state.last_paid_height,
            consecutive_payments: protx.state.consecutive_payments,
            pose_penalty: protx.state.pose_penalty,
            pose_revived_height: protx.state.pose_revived_height,
            pose_ban_height: protx.state.pose_ban_height,
            revocation_reason: protx.state.revocation_reason,
            owner_address: protx.state.owner_address.clone(),
            voting_address: protx.state.voting_address.clone(),
            payout_address: protx.state.payout_address.clone(),
            pub_key_operator: protx.state.pub_key_operator.clone(),
            platform_node_id: protx.state.platform_node_id.clone(),
            platform_http_port: protx.state.platform_http_port,
            platform_p2p_port: protx.state.platform_p2p_port,
            is_platform_banned: protx.meta_info.is_platform_banned,
            confirmations: protx.confirmations,
            last_dsq: protx.meta_info.last_dsq,
            mixing_tx_count: protx.meta_info.mixing_tx_count,
        }
    }
}
